Make X11 backend and Xwayland support optional

cl91 opened the pull request:
This makes the X11 backend and Xwayland support optional at compile time --- if relevant X11 and XCB packages are found, then they are enabled. If not, they are automatically disabled.

This would be useful for systems that don't need X11 and don't want any X11 dependencies (for instance, embedded systems). I'm currently building a Wayland-only Gentoo system (that is, no X11 dependencies at all), and have successfully tested sway with this commit (In fact, I'm writing this under sway and midori right now. All built without any X11 dependencies (libX11 and friends).

Cloudef commented:
https://github.com/Cloudef/wlc/issues/47 The correct way is to do this issue.

I don't generally like how many #ifdef paths this patch introduces, especially in places that are not directly related to X11 or Xwayland.

If you can reduce that clutter by no-op:ing datastructures and their access functions instead of disabling code, I would accept this.

E.g. Instead of doing 
```
+#ifdef ENABLE_XWAYLAND
     wlc_xwm_release(&compositor->xwm);
 +#endif
```

Do the #ifdef inside the wlc_xwm_release. The xwm basically would be no-op datastructure when it's not compiled in.

cl91 commented:
I have refactored the code to remove the #ifdef's and added no-ops and dummy data structures for the case where xwayland is disabled. It should look much cleaner now.

Cloudef commented on src/compositor/view.h:
```diff
@@ -59,6 +59,10 @@ struct wlc_view {
    } state;
 };
 
+static inline bool is_x11_view(struct wlc_view *view) {
```
Put storage class and return value in own line e.g:
```
static inline bool
is_x11_view(...)
{
   ...
}
```

This is the style wlc uses.

Cloudef commented on include/wlc/defines.h:
```diff
@@ -12,11 +12,13 @@ extern "C" {
 #  define WLC_NONULLV(...) __attribute__((nonnull(__VA_ARGS__)))
 #  define WLC_PURE __attribute__((pure))
 #  define WLC_CONST __attribute__((const))
+#  define WLC_UNUSED __attribute__((unused))
```
Instead of __attribute__((unused)) wlc has used to (void)foobar; to ignore "not used" warnings.

Cloudef commented on src/xwayland/xwm.h:
```diff
@@ -36,4 +53,43 @@ WLC_NONULL void wlc_x11_window_close(struct wlc_x11_window *win);
+WLC_NONULL static inline void wlc_x11_window_configure(struct wlc_x11_window *win WLC_UNUSED, const struct wlc_geometry *g WLC_UNUSED) {}
+WLC_NONULL static inline void wlc_x11_window_set_state(struct wlc_x11_window *win WLC_UNUSED, enum wlc_view_state_bit state WLC_UNUSED, bool toggle WLC_UNUSED) {}
+
+WLC_NONULL static inline bool wlc_x11_window_set_active(struct wlc_x11_window *win WLC_UNUSED, bool active WLC_UNUSED)
```
Remove WLC_UNUSED and put `(void)win; (void)active;` inside function body.

Cloudef commented:
Yup it's much better contained now. Fix the new comments and then it should be ok for squash and merge.

cl91 commented:
I have fixed the style issues with the new commits now.

Cloudef commented:
Thanks, it looks good. Squash and I will merge it :)

cl91 commented:
Done.

Cloudef commented:
Thanks, will smoke the builds later today.
